decoupling recorder, observer and pattern matcher

Sczlog opened the pull request:
refactor code structure for the sake of further development.
now recorder have three module:
1. observer:
A module spys on the user interaction, it will listen event and emit it to recorder.
2. matcher:
A module processs sending event and tranform them into a user action, then send it back to recorder.
3. recorder:
A module organizes observers and matcher, a recorder can have many observers and one matcher, it will provide a bridge between observers and matcher, and communicate with outer.

SimpleRecorder is an alternative to the previous recorder.

Sczlog committed:
decoupling recorder, observer and match pattern.

Sczlog committed:
some misc update

1. not to pre process step event in recorder.
2. clean events and target when stop the pattern matcher.

Sczlog committed:
update yarn.lock to public registry

Sczlog committed:
use enum to replace plain text for matcher's event key。

netweng commented on src/index.ts:
```diff
@@ -1,130 +1,251 @@
-      this.eventHandler.bind(this),
-    );
-    this.metaQuerier = metaQuerier || new BasicMetaQuerier();
+    this._recorder.extendAction<PatternMatcherExtendParams>({
```
maybe we can have `buiid-in-action` files to manage these code

Sczlog committed:
init test environment

update tsconfig
init test file for modules.

Sczlog committed:
Merge branch 'webzard-io:master' into dev

netweng commented on src/matcher.ts:
```diff
@@ -0,0 +1,168 @@
+    );
+  }
+
+  private handleNewEvent(stepEvent: StepEvent, target: HTMLElement | null) {
```
which situation would be `continue`? which would be `return`?

Sczlog replied:
the handler should return `'return'|'emit'|'collect'|undefined`
`'return'` mean we should ignore this event and not process it.
`'undefined'` mean we should ignore current handler but keep processing it.
`'emit|collect'` mean we should do some action and stop current handler group and continue processing the event.

netweng replied:
can we replace preStep, collectStep, postStep  by using:
``` typescript
shouldEmitBeforeCollectStep: () => boolean;
shouldCollect: () => boolean;
shouldEmitAfterCollectStep: () => boolean;
```

Sczlog replied:
I think boolean is not clear in this situation as the return value has 3 state, using enum or string will be better.
Name doesn't matter, I will change it.

netweng replied:
boolean | undefined ?

netweng commented on src/recorder.ts:
```diff
@@ -0,0 +1,130 @@
+      );
+      return observer;
+    }
+    if (this.emitterMap.has(observer)) {
```
we can have only one listener on an event ?

Sczlog replied:
It means one observer should only have a single group processing method.
When extended, they are immutable.
If you want to modify the action, you should firstly suspend the the recorder and remove this action and re-extend the observer.

netweng commented on src/observers.ts:
```diff
@@ -62,12 +49,15 @@ export class EventObserver {
   private throttleManager = new ThrottleManager();
 
-  constructor(win: Window, doc: Document, onEmit: EmitHandler) {
+  constructor(win: Window, doc: Document) {
```
we still need doc?

Sczlog replied:
We still need document element for scroll event,but we can use `window.document` to replace it.

Sczlog committed:
bug fix and unit test add

1. make an observer can be extends to multiple recorder.
2. create unit test for observer and matcher.
3. fix some bug found when create unit test.

Sczlog committed:
update unit test

1. update unit test for observer.
2. fix wheel observer doesn't check state of observer.

Sczlog committed:
Merge branch 'dev' of https://github.com/Sczlog/interaction-recorder into dev

netweng commented on __tests__/units/matcher.spec.ts:
```diff
@@ -0,0 +1,289 @@
+import { getMockedObserver } from '../util/mock';
+let matcher: PatternMatcher;
+
+const buildParams = (
```
can we import type from  matcher.ts

netweng commented on __tests__/units/observer.spec.ts:
```diff
@@ -0,0 +1,604 @@
+      emitSpy.calls.reset();
+    });
+
+    describe('observe mouse interaction', () => {
```
should we need test every event handler?
maybe we just need to test the functions, but event observer should be test in puppeteer

Sczlog committed:
update unit test for builtin recorder

1. update unit test for buildin recorder
2. use enum as pattern matcher's step control result;

Sczlog committed:
Update jest.config.js

Sczlog committed:
remove unit test for current pr

netweng commented:
LGTM
